client/www/components/dash/AppAuth.tsx

```diff
@@ -720,11 +735,544 @@ function GoogleClients({
+// -----
+
+function AddClerkProviderForm({
+  app,
+  onAddProvider,
+}: {
+  app: InstantApp;
+  onAddProvider: (provider: OAuthServiceProvider) => void;
+}) {
+  const token = useContext(TokenContext);
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+  const addClerkProvider = async () => {
+    setIsLoading(true);
+    try {
+      const resp = await addProvider({
+        token,
+        appId: app.id,
+        providerName: 'clerk',
+      });
+      onAddProvider(resp.provider);
+    } catch (e) {
+      console.error(e);
+      const msg =
+        messageFromInstantError(e as InstantError) ||
+        'There was an error setting up Clerk.';
+      errorToast(msg, { autoClose: 5000 });
+      // report error
+    } finally {
+      setIsLoading(false);
+    }
+  };
+  return (
+    <div>
+      <Button
+        loading={isLoading}
+        variant="secondary"
+        onClick={addClerkProvider}
+      >
+        <span className="flex items-center space-x-2">
+          <Image src={clerkLogoSvg} />
+          <span>Setup Clerk</span>
+        </span>
+      </Button>
+    </div>
+  );
+}
+
+// Base64 decode, switching to url-safe decode if we hit an error
+// Can't be sure which method Clerk uses because you can't generate
+// `+` or `/` with characters that go in a normal host. Urls with
+// chinese characters exist, they might encode to `+` or `/`, and
+// Clerk might support them, so we'll be safe and do both.
+function base64Decode(s: string) {
+  try {
+    return Buffer.from(s, 'base64').toString('utf-8');
+  } catch (e) {
+    return Buffer.from(s, 'base64url').toString('utf-8');
+  }
+}
+
+function domainFromClerkKey(key: string): string | null {
+  try {
+    const parts = key.split('_');
+    const domainPartB64 = parts[parts.length - 1];
+    const domainPart = base64Decode(domainPartB64);
+    return domainPart.replace('$', '');
+  } catch (e) {
+    console.error('Error getting domain from clerk key', e);
+    return null;
+  }
+}
+
+function clerkExampleCode({
+  appId,
+  clientName,
+  clerkPublishableKey,
+}: {
+  appId: string;
+  clientName: string;
+  clerkPublishableKey: string;
+}) {
+  return /* replace-me-with-js-to-format */ `import {
+  useAuth,
+  ClerkProvider,
+  SignInButton,
+  SignedIn,
+  SignedOut,
+} from "@clerk/clerk-react";
+import { init } from "@instantdb/react";
+import { useEffect } from "react";
+
+const db = init({ appId: "${appId}" });
+
+function ClerkSignedInComponent() {
+  const { getToken, signOut } = useAuth();
+
+  const signInToInstantWithClerkToken = async () => {
+    // getToken gets the jwt from Clerk for your signed in user.
+    const idToken = await getToken();
+
+    if (!idToken) {
+      // No jwt, can't sign in to instant
+      return;
+    }
+
+    // Create a long-lived session with Instant for your clerk user
+    // It will look up the user by email or create a new user with
+    // the email address in the session token.
+    db.auth.signInWithIdToken({
```

Comment:
Thought: 

With this `useEffect`, afaik `signInWithIdToken` would end up running on each page load, even if we already authed before. 

I worry this would cause some thrash in Reactor, because we would always send an extra ['user changed' update](https://github.com/instantdb/instant/blob/main/client/packages/core/src/Reactor.js#L1120C10-L1120C27)
 
One option to I could think of, is to show a demo where we do the login, as part of Clerk's redirect. Here's an example I wrote for some users: 

https://github.com/stopachka/clerk_auth_example/blob/main/frontend/src/App.tsx#L31-L35

Another could be to make `signInWithIdToken` smarter, though I am not 100% sure how we could achieve it. 

What do you think @dwwoelfel ?

Reply:
Resolved with https://github.com/instantdb/instant/pull/97/commits/e9d98ad7a4ed0b9190a3084a0327fed84df9d995

We'll reuse the user's existing token if it's still valid and belongs to the same user. Then in `changeCurrentUser`, we'll exit early if the current user is the same as the new user.